File: DoKey.Core/KeysEngine.fs
namespace DoKey.Core

module KeysEngine = 

    open Domain
    open DomainOperations

    let EmptyMappedKey = { send = ""; trigger = ""; isCaps = false } 

    let GetMappedKey isCaps (trigger:string) keys =
        let predicate (mappedKey:MappedKey) = mappedKey.trigger = trigger.ToLower() && mappedKey.isCaps = isCaps
        let result = Seq.tryFind predicate keys
        match result with
            | Some i -> i
            | None -> EmptyMappedKey 

    let IsDownFirstStep firstStep key =
        firstStep = "" && IsTwoStep key

    let GetMappedKeyNormal firstStep key keys =
        let isDownFirstStep = IsDownFirstStep firstStep key
        match isDownFirstStep with
            | true-> EmptyMappedKey 
            | _ -> GetMappedKey false (firstStep + key) keys 
    
    let GetMappedKeyNormalAndInsertWithCapital appState key keys =
        match appState.state with 
            | State.Off -> EmptyMappedKey
            | _ -> match appState.modificators.caps with
                   | false -> EmptyMappedKey
                   | _ -> GetMappedKey true (appState.firstStep + key) keys 
        
    let GetNextAppStateByStateChange key appState =
        let nextState = Domain.GetNextStateByKey key appState.state
        match nextState with
            | None -> None
            | _-> Some {state = nextState.Value; modificators = appState.modificators; firstStep=""; preventEscOnCapsUp = true}

    let GetNextAppStateByESC inputKey appState = 
        match inputKey.isEsc with
            | false -> None
            | _ -> match appState.state with
                    | State.Insert -> Some {state = GetPrevState appState.state; modificators = appState.modificators; 
                        firstStep=""; preventEscOnCapsUp = appState.preventEscOnCapsUp}
                    | _ -> None


    let CanProcessNormalMode appState =
        match appState.state with
            | State.Normal -> match appState.modificators.win with
                                | true -> false 
                                | _ -> true
            | _ -> false


    let ProcessNormalMode appState inputKey keys =
        let isDownFirstStep = appState .firstStep = "" && IsTwoStep inputKey.key
        let firstStepNext = if isDownFirstStep then inputKey.key else ""
        let sendDoKey = GetMappedKeyNormal appState.firstStep inputKey.key keys
        let preventKeyProcess = inputKey.isLetterOrDigit || sendDoKey.send <> ""
        let nextAppState = {state = appState.state; modificators = appState.modificators; firstStep = firstStepNext; preventEscOnCapsUp = appState.preventEscOnCapsUp}
        {appState = nextAppState ; send = sendDoKey.send ; preventKeyProcess = preventKeyProcess }
            
    let ProcessKey appState inputKey keys=
        if CanProcessNormalMode appState 
        then ProcessNormalMode appState inputKey keys
        else {appState = appState; send = ""; preventKeyProcess = false}

    //let ProcessNormalMode appState inputKey =
    //    match CanProcessNormalMode appState with
    //        | false -> ProcessKey
    //        | _ ->

    //private KeysEngineResult ProcessNormalMode()
        //{

        //  if (AppState.state != State.Normal) return null;
        //  if (AppState.modificators.win) return null;

        //  var isDownFirstStep = AppState.firstStep == "" && DomainOperations.IsTwoStep(inputKey.key);

        //  var firstStepNext = isDownFirstStep ? inputKey.key : "";

        //  var sendDoKey = DoKey.Core.KeysEngine.GetMappedKeyNormal(AppState.firstStep, inputKey.key.ToString(), config.mappedKeys);

        //  var preventKeyProcess = inputKey.isLetterOrDigit || sendDoKey.send != "";

        //  return new KeysEngineResult(new AppState(this.AppState.state, this.AppState.modificators, firstStepNext, this.AppState.preventEscOnCapsUp), sendDoKey.send, preventKeyProcess);

        //}

File: DoKey.App/KeysEngine.cs
using DoKey.Core;
using static DoKey.Core.Domain;

namespace DoKey
{

  public class KeysEngine
  {

    public Config config { get; set; }
    public AppState AppStateX { get; set; }
    public InputKey inputKey { get; set; }
    public bool isUp { get; set; }

    public KeysEngineResult ProcessKey()
    {

      if (inputKey.isCaps) return ProcessCapital();
      if (inputKey.isModif) return ProcessModificators();

      if (isUp) return new KeysEngineResult(AppStateX, "", false);

      //------------------
      var appState = ProcessStateChange();
      if (appState != null) return DomainOperations.CreateEngineResultChangeAppState(appState);
      //------------------------

      if (AppStateX.state == State.Off) return new KeysEngineResult(AppStateX, "", false);

      var output = ProcessNormalAndInsertWithCapital();
      if (output != null) return output;

      return Core.KeysEngine.ProcessKey(AppStateX, inputKey, config.mappedKeys);

    }

    private AppState ProcessStateChange()
    {
      if (AppStateX.modificators.caps)
      {
        var x = Core.KeysEngine.GetNextAppStateByStateChange(inputKey.key, AppStateX);
        try        {          return x.Value;        }
        catch        {          return null;        }
      }
      else
      {
        var x = Core.KeysEngine.GetNextAppStateByESC(inputKey, AppStateX);
        try { return x.Value; }
        catch { return null; }
      }
    }

    private KeysEngineResult ProcessModificators()
    {

      var modificators = ModificatorsOperations.GetNextModificators(AppStateX.modificators, inputKey, isUp);

      return new KeysEngineResult(new AppState(AppStateX.state, modificators, AppStateX.firstStep, AppStateX.preventEscOnCapsUp), "", false);

    }

    private KeysEngineResult ProcessCapital()
    {

      var sendKeys = "";
      var preventEscOnNextCapitalUp = AppStateX.preventEscOnCapsUp;

      if (isUp)
      {
        if (AppStateX.preventEscOnCapsUp)
        {
          preventEscOnNextCapitalUp = false;
        }
        else
        {
          sendKeys = "{ESC}";
        }
      }

      var modif = ModificatorsOperations.GetNextModificators(AppStateX.modificators, inputKey, isUp);

      return new KeysEngineResult(new AppState(AppStateX.state, modif, "", preventEscOnNextCapitalUp), sendKeys, true);

    }

    private KeysEngineResult ProcessNormalAndInsertWithCapital()
    {

      var sendKeys = DoKey.Core.KeysEngine.GetMappedKeyNormalAndInsertWithCapital(AppStateX, inputKey.key.ToString(), config.mappedKeys);

      if (sendKeys.send == "") return null;

      return new KeysEngineResult(new AppState(AppStateX.state, AppStateX.modificators, "", true), sendKeys.send, true);

    }

  }
}

